src/Standards/PSR12/Sniffs/Functions/NullableTypeDeclarationSniff.php

```diff
@@ -0,0 +1,73 @@
+<?php
+/**
+ * Verifies that nullable typehints are lacking superfluous whitespace, e.g. ?int
+ *
+ * @author    Greg Sherwood <gsherwood@squiz.net>
+ * @copyright 2006-2015 Squiz Pty Ltd (ABN 77 084 670 600)
+ * @license   https://github.com/squizlabs/PHP_CodeSniffer/blob/master/licence.txt BSD Licence
+ */
+
+namespace PHP_CodeSniffer\Standards\PSR12\Sniffs\Functions;
+
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Sniffs\Sniff;
+use PHP_CodeSniffer\Util\Tokens;
+
+class NullableTypeDeclarationSniff implements Sniff
+{
+
+
+    /**
+     * Returns an array of tokens this test wants to listen for.
+     *
+     * @return array
+     */
+    public function register()
+    {
+        return [T_NULLABLE];
+
+    }//end register()
+
+
+    /**
+     * Processes this test, when one of its tokens is encountered.
+     *
+     * @param \PHP_CodeSniffer\Files\File $phpcsFile The file being scanned.
+     * @param int                         $stackPtr  The position of the current token in the
+     *                                               stack passed in $tokens.
+     *
+     * @return void
+     */
+    public function process(File $phpcsFile, $stackPtr)
+    {
+        $nextValidTokenPtr = $phpcsFile->findNext([T_STRING, T_NS_SEPARATOR], ($stackPtr + 1), null);
```

Comment:
Further on my previous comment about this line: https://github.com/squizlabs/PHP_CodeSniffer/pull/2022#discussion_r191965848

This still will not work as intended.

Additional test cases which should be added and demonstrate the problem:
```php
function testSelf( ? self $self ) : ? self {}
function testParent( ? parent $parent ) : ? parent {}
function testCallable( ? callable $callable ) : ? callable {}
```

Reply:
I've added the additional test cases and the issue appears not to be the reverse check - the additional cases *are* marked as violations - but not taking into account all the allowed/expected type hints and therefor not marking them as fixable. 

What I did is restructured the code to the reverse check and cleaned it up a bit so the maximum nesting depth of conditionals is two. I added three additional tokens `T_CALLABLE`, `T_SELF` and `T_PARENT`. I tried to find an exhaustive list of allowed type hint tokens to make sure I'm not forgetting any tokens. 

I also updated the copyright comment.

Reply:
> I tried to find an exhaustive list of allowed type hint tokens to make sure I'm not forgetting any tokens.

Looking at the current list, I'm pretty sure you're ok now.